simplify-parents: add a command to remove redundant parents

torquestomp opened the pull request:
<!--
There's no need to add anything here, but feel free to add a personal message.
Please describe the changes in this PR in the commit message(s) instead, with
each commit representing one logical change. Address code review comments by
rewriting the commits rather than adding commits on top. Use force-push when
pushing the updated commits (`jj git push` does that automatically when you
rewrite commits). Merge the PR at will once it's been approved. See
https://github.com/martinvonz/jj/blob/main/docs/contributing.md for details.
Note that you need to sign Google's CLA to contribute.
-->

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [x] I have added tests to cover my changes

samueltardieu commented on cli/src/commands/simplify.rs:
```diff
@@ -0,0 +1,133 @@
+
+    tx.repo_mut()
+        .rewrite_commit(settings, commit)
+        .set_parents(new_heads.into_iter().collect())
```
I think you might reorder the untouched parents in an arbitrary order. Order should probably be preserved.

samueltardieu commented on cli/tests/cli-reference@.md.snap:
```diff
@@ -1811,6 +1813,23 @@ Show commit description and changes in a revision
+
+Removes any and all redundant parent edges going into any of the specified revisions. By definition, this has no effect on any commit contents and should not cause any working copy changes or conflicts.
+
+Immutable commits are automatically filtered from the specified inputs.
```
Shouldn't `--ignore-immutable` be honored instead?

samueltardieu commented on cli/tests/cli-reference@.md.snap:
```diff
@@ -1811,6 +1813,23 @@ Show commit description and changes in a revision
+
+Simplify parent edges for the specified revision(s).
+
+Removes any and all redundant parent edges going into any of the specified revisions. By definition, this has no effect on any commit contents and should not cause any working copy changes or conflicts.
```
nit: "all" covers "any and all"

samueltardieu commented on cli/tests/cli-reference@.md.snap:
```diff
@@ -1811,6 +1813,23 @@ Show commit description and changes in a revision
+
+Simplify parent edges for the specified revision(s).
+
+Removes any and all redundant parent edges going into any of the specified revisions. By definition, this has no effect on any commit contents and should not cause any working copy changes or conflicts.
```
It was not clear to me what a redundant parent edge was before reading the code and playing with the code. Maybe a textual example would be good ("For example, a commit with multiple parents descending from one another will only depend on its closest ancestors." or something like that, even though it is clear what "closest ancestors" mean here).

torquestomp replied:
Added a paragraph here.

samueltardieu commented:
I don't want to approve this change myself because I've not followed closely enough the discussions to know the policy about adding new commands. In particular, is "simplify" clear enough, or should it be "simplify-tree", or whatever?

But I would have had a use for it as recently as today, when one of my PR (change A) was merged into `main` today: one of my development subtree was based both on this change A and another unmerged one (change B). After rebasing B onto main, I ended up for my dev subtree parents with a merge between a descendant of main (B) and what was now a parent of main (A). A `jj simplify` here would have saved me the manual rebasing of my subtree on B only.

In this case it was only one subtree with only two parents, and a simple operation, but with a more complex development tree I can see how it would have been even more useful.

yuja commented on cli/src/commands/simplify.rs:
```diff
@@ -0,0 +1,139 @@
+    let mut stats = SimplifyStats::default();
+    for commit in commits {
+        stats.add(simplify_commit(command.settings(), &mut tx, &commit)?);
+    }
```
Perhaps, we can use `repo.transform_descendants()` to rewrite commits in topological order. It will save interim predecessor commits.

Maybe we can also build some cache for fast heads computation, but I'm not sure if doing that makes sense. If the history is mostly linear, `index.heads()` might be fast enough.

torquestomp replied:
This also lets me re-use the 'simplify_ancestor_merge()' function

yuja commented on cli/src/commands/simplify.rs:
```diff
@@ -0,0 +1,139 @@
+            )?;
+        }
+    }
+    tx.finish(ui, format!("simplified {} commits", orig_commits))?;
```
nit: "simplify .."

yuja reviewed (commented):
> In particular, is "simplify" clear enough, or should it be "simplify-tree", or whatever?

I don't have a good suggestion, but I feel "simplify" is too generic. If we use "simplify-<something>", "simplify-parents" is probably better.

torquestomp commented:
> > In particular, is "simplify" clear enough, or should it be "simplify-tree", or whatever?
> 
> I don't have a good suggestion, but I feel "simplify" is too generic. If we use "simplify-", "simplify-parents" is probably better.

Done, changed to simplify-parents for now.

Idk if the file should still be named 'simplify.rs' to leave room for others, but I changed it to simplify_parents.rs also for now

thoughtpolice commented on CHANGELOG.md:
```diff
@@ -64,6 +64,8 @@ to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 * CommitId / ChangeId template types now support `.normal_hex()`.
 
+* New command `jj simplify` will remove redundant parent edges.
```
`simplify-parents`

samueltardieu reviewed (commented):
nit: the commit message still contains `simplify` instead of `simplify-parents`

torquestomp commented:
> nit: the commit message still contains `simplify` instead of `simplify-parents`

Find and replace, my greatest weakness

yuja commented on cli/src/commands/simplify_parents.rs:
```diff
@@ -0,0 +1,119 @@
+    if num_new_heads < num_old_heads {
+        rewriter.reparent(settings)?.write()?;
+
+        let new_stats = SimplifyStats::for_commit(num_old_heads - num_new_heads);
```
nit: maybe simpler to just do `stats.commits += 1; ...`?

yuja commented on cli/src/commands/simplify_parents.rs:
```diff
@@ -0,0 +1,119 @@
+    rewriter.simplify_ancestor_merge();
+    let num_new_heads = rewriter.new_parents().len();
+
+    if num_new_heads < num_old_heads {
+        rewriter.reparent(settings)?.write()?;
```
probably better to do `.reparent()` if `.parents_changed()` so the descendants won't be rebased later by `tx.finish()`.

yuja commented:
> `simplify-parents` seems rather wordy to be in the top-level namespace, but other than that I think it's fine. I suspect it might be best to go under a `utils` subcommand or something like we've talked about before, but I'm interested in what other people think.

Yeah. If we had a generic graph reshaping command or command namespace, `simplify` can be moved there. This could also be `rebase --simplify-ancestor-merge`, but it's inconvenient if the user case is to eliminate redundant edges created by autorebase.

@ilyagr any thoughts?

ilyagr commented:
I think `simplify-parents` is fine for now, unless we get a better idea.

One alternative I can think of is to have a subcommand, e.g. `jj graph simplify-parents` or `jj graph-util simplify-parents`. 

Yet another option would be to make `jj rebase` a namespace with `jj rebase source` instead of `jj rebase -s`, and then we could have `jj rebase simplify-parents`.

Another command that could go into this namespace (or be an option to rebase) is `jj reparent-children --from X --to Y` (or `move-children`) that replaces `X` with `Y` in the list of any commit's children (without breaking merge commits).

torquestomp commented:
I like the idea of `jj graph simplify-parents`, it would probably make sense to also move `jj parallelize` into this subgroup?  (Also is anyone making `jj linearize` yet, that would also fit here?)

yuja commented:
Operation history is also a "graph", so I think `jj graph` is still vague.

> it would probably make sense to also move `jj parallelize` into this subgroup?

It might make sense, but I tend to prefer flattened command structure. `jj parallelize` isn't so ambiguous compared to `jj simplify` (without `-parents`/`-merge`.)

martinvonz commented:
If we add a `graph` command category, then it seems like even `rebase` would belong under it, but if we do that, then basically everyone would create an alias for it anyway.

ilyagr commented:
> Operation history is also a "graph", so I think jj graph is still vague.

This is a good point. `jj commits` or `jj commitgraph` or `jj cgraph` seem like obvious alternatives, but I'm not sure they are great. Or we could have `jj graph`, `jj evo graph` (or `evo-graph` or `evograph`), and `jj op graph`.

> If we add a graph command category, then it seems like even rebase would belong under it, but if we do that, then basically everyone would create an alias for it anyway.

I wouldn't mind `jj rebase` being an alias for `jj graph rebase`. On the other hand, if we just want the uncommon commands to be in a subcommand, it could be `jj graph-utils simplify-parents` or `jj util graph simplify-parents`.

----

Another idea, based on Yuya's comment is `jj simplify parents`. But, in any case, we could start with `jj simplify-parents`.

ilyagr commented on cli/src/commands/simplify_parents.rs:
```diff
@@ -0,0 +1,119 @@
+
+/// Simplify parent edges for the specified revision(s).
+///
+/// Removes all redundant parent edges going into any of the specified
```
(Optional) The following seems a bit clearer to me for the first sentence:

> Removes any parents of each of the the specified revisions that are also indirect ancestors of the same revision

For the second sentence, I'd also be more decisive:

> This has no effect on any revision's contents or on the working copy.

I'm also tempted to write "This should have no effect on any...", and I'm OK with that if you prefer it, but I think style guides or Orwell's "Politics and the English language" would suggest dropping the "should"

torquestomp replied:
Rewored this mostly according to the suggestions.

ilyagr commented on cli/src/commands/simplify_parents.rs:
```diff
@@ -0,0 +1,119 @@
+    /// Simplify specified revision(s) together with their trees of descendants
+    /// (can be repeated)
+    #[arg(long, short, group = "revision-args")]
+    sources: Vec<RevisionArg>,
```
I think we use singular "source" in `rebase`, and that singular is also better (since to the user it looks like `--source a --source b`, where `a` and `b` usually expand to a single revision).

ilyagr commented on cli/src/commands/simplify_parents.rs:
```diff
@@ -0,0 +1,119 @@
+    let mut workspace_command = command.workspace_helper(ui)?;
+    let revs = RevsetExpression::descendants(
+        workspace_command
+            .parse_union_revsets(ui, &args.sources)?
```
Relatedly to the previous comment, I think this should use `resolve_some_revsets_default_single`.

yuja replied:
A bit off topic, but do you remember why `rebase -s`/`-b` requires `all:`? I don't think `-s multiple-roots` is a footgun. `-s`/`-b` are similar to `-r`.

ilyagr replied:
`-s multiple-roots` can be a footgun, at least if it acts in the way that makes sense to me.

I'd like `-s A -s B -d X` to be equivalent to `-s all:A|B -d X` and result in both `A` and `B` becoming direct children of `X`, *independently* of whether one of `A` or `B` is a descendant of the other one.

If so, `rebase -s A::` would give a very confusing result, and so I'd prefer that gave an error, and the user did `rebase -s all:A::` if that's what they wanted.

There might be other ways to negotiate this problem, which would involve changing what `-s multiple_revision` means (e.g. automatically rewrite it to `-s roots(multiple_revisions)`, making `-s A|B` result in different thing if `B` is a child of `A`). However, the behavior I described seems sensible to me and seems to require `all:`.

ilyagr replied:
Though, this makes me think that all of my arguments for `rebase` do not apply to `simplify-parents`. So, maybe we could keep using `parse_union_revset` here. I'm not sure whether or not this slight extra flexibility is worth making the UI subtly different from `rebase`'s.

I think, for `simplify-parents`, I still mildly prefer making the UI the same as `rebase`'s, but I no longer think it's particularly important.

yuja replied:
> I'd like `-s A -s B -d X` to be equivalent to `-s all:A|B -d X` and result in both `A` and `B` becoming direct children of `X`, _independently_ of whether one of `A` or `B` is a descendant of the other one.

Ah, good point. For `-s A|B`, the user might expect `roots(A|B)` to be the destination children (if `-s` accepted revision ranges by default.)

> Though, this makes me think that all of my arguments for `rebase` do not apply to `simplify-parents`.

Yes. `simplify-parents` is similar to `jj fix -s`.

ilyagr replied:
>  For -s A|B, the user might expect roots(A|B) to be the destination children

Yes, or `roots(connected(A|B)`. Not having to choose between these slightly different behaviors is one advantage of the current behavior.

ilyagr reviewed (commented):
For now, here are a few bikesheddy comments. I hope they are uncontroversial, but I thought I'd send them out early in just in case there are other opinions.

ilyagr commented on cli/src/commands/simplify_parents.rs:
```diff
@@ -0,0 +1,119 @@
+    )
+    .union(
+        workspace_command
+            .parse_union_revsets(ui, &args.revisions)?
```
(No action needed) For "revisions", `rebase` currently uses plural and `parse_union_revsets`, so I think things are correct as they are.

ilyagr commented:
By the way, I agree with this approach of dealing with redundant ancestry overall. It seems more predictable to me than trying to do simplification automatically in some situations. (Well, I suppose, if we have some great ideas about auto-simplification, this PR doesn't prevent us trying to do that too, but I'm not optimistic about that at the moment).

ilyagr commented on cli/tests/test_simplify_parents_command.rs:
```diff
@@ -0,0 +1,142 @@
+    ├─╯
+    ○  a
+    ◆
+    "###);
```
Another interesting case to me (that I think we should test) is `jj simplify-parents -r 'all()'` for something like

```
    @    d
    ├─╮
    ○ │  c
    │ ○  b
    ├─╯
    ○  a
    ◆
```

I think it should be a no-op, and I haven't read the code carefully enough to check whether it will be.

ilyagr replied:
Thanks for adding! I hadn't noticed when you did it. I'll hopefully take one last look at this soon.

ilyagr commented on CHANGELOG.md:
```diff
@@ -101,6 +101,16 @@ to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+    from leaving the working copy in a stale state.
+ 
+  * `jj git fetch -b <remote-git-branch-name>` will now warn if the branch(es)
+>>>>>>>
```
There's a forgotten conflict marker here.

samueltardieu commented on CHANGELOG.md:
```diff
@@ -101,6 +101,16 @@ to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 * `jj git fetch -b <remote-git-branch-name>` will now warn if the branch(es)
+
+* New command `jj simplify-parents` will remove redundant parent edges.
```
It looks like this will be inserted in the 0.22 release section, this should be moved up to the unreleased section.

ilyagr reviewed (approved):
This all LGTM. The new clippy warning should be easy to fix. Thank you!

bnjmnt4n commented:
Would it be a good idea to set a default revset for `jj simplify-parents` (like [`jj fix`](https://github.com/martinvonz/jj/pull/3866))?
